ethereum/contracts/IncentivizedInboundChannel.sol

```diff
@@ -0,0 +1,117 @@
+
+    RewardSource private rewardSource;
+
+    BeefyLightClient public beefyLightClient;
+
+    constructor(BeefyLightClient _beefyLightClient) {
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        beefyLightClient = _beefyLightClient;
+        nonce = 0;
+    }
+
+    // Once-off post-construction call to set initial configuration.
+    function initialize(address _configUpdater, address _rewardSource)
+        external
+    {
+        require(
+            hasRole(DEFAULT_ADMIN_ROLE, msg.sender),
+            "Caller is unauthorized"
+        );
+
+        // Set initial configuration
+        grantRole(CONFIG_UPDATE_ROLE, _configUpdater);
+        rewardSource = RewardSource(_rewardSource);
+
+        // drop admin privileges
+        renounceRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    function submit(
+        Message[] calldata _messages,
+        ParachainLightClient.ParachainVerifyInput
+            calldata _parachainVerifyInput,
+        ParachainLightClient.BeefyMMRLeafPartial calldata _beefyMMRLeafPartial,
+        uint256 _beefyMMRLeafIndex,
+        uint256 _beefyMMRLeafCount,
+        bytes32[] calldata _beefyMMRLeafProof
+    ) public {
+        // Proof
+        // 1. Compute our parachain's message `commitment` by ABI encoding and hashing the `_messages`
+        bytes32 commitment = keccak256(abi.encode(_messages));
+
+        ParachainLightClient.verifyCommitmentInParachain(
+            commitment,
+            _parachainVerifyInput,
+            _beefyMMRLeafPartial,
+            _beefyMMRLeafIndex,
+            _beefyMMRLeafCount,
+            _beefyMMRLeafProof,
+            beefyLightClient
+        );
+
+        // Require there is enough gas to play all messages
+        require(
+            gasleft() >= (_messages.length * MAX_GAS_PER_MESSAGE) + GAS_BUFFER,
+            "insufficient gas for delivery of all messages"
+        );
+
+        processMessages(payable(msg.sender), _messages);
+    }
+
+    function processMessages(
+        address payable _relayer,
+        Message[] calldata _messages
+    ) internal {
+        uint256 _rewardAmount = 0;
+
+        for (uint256 i = 0; i < _messages.length; i++) {
+            // Check message nonce is correct and increment nonce for replay protection
+            require(_messages[i].nonce == nonce + 1, "invalid nonce");
+
+            nonce = nonce + 1;
+
+            // Deliver the message to the target
+            // Delivery will have fixed maximum gas allowed for the target app
+            (bool success, ) = _messages[i].target.call{
+                value: 0,
+                gas: MAX_GAS_PER_MESSAGE
+            }(_messages[i].payload);
+
+            _rewardAmount = _rewardAmount + _messages[i].fee;
+            emit MessageDispatched(_messages[i].nonce, success);
+        }
+
+        // Attempt to reward the relayer
+        try rewardSource.reward(_relayer, _rewardAmount) {} catch {
```

Comment:
Will fail if messages are received before initialize is called. Might be worthwhile to revert in that case (if _rewardSource is zero).

Reply:
yeh maybe we should just remove the try/catch and let it revert?

Reply:
The reason for the try/catch around this call was that reward source needs to be manually filled from fees collected on Substrate, right @vgeddes ?

Reply:
The intention was to add some safety in case for example the reward fails because there are not enough funds in the rewardSource to pay the reward. If that were to happen, the bridge would potentially get blocked. This could only occur if the bridge was already broken though, so I think fine to revert it.

Additionally, even if somehow this did happen, we could refill the rewardSource manually so I think this is fine to remove.

Reply:
Agree. 

As you said, the rewardSource can’t be empty without a serious incident with the bridge - we burn SnowETH to pay for the transaction on the sending side, so it needs to be locked here.